### PR TITLE
ESDT improvements chain simulator tests and fixes

### PR DESCRIPTION
**Pull request by ssd04**
## Reasoning behind the pull request
- Reference vm common fix
- Integration tests with chain simulator

## Testing procedure
- Standard system test

## Pre-requisites

Based on the [Contributing Guidelines](https://github.com/multiversx/mx-chain-go/blob/master/.github/CONTRIBUTING.md#branches-management) the PR author and the reviewers must check the following requirements are met:
- was the PR targeted to the correct branch?
- if this is a larger feature that probably needs more than one PR, is there a `feat` branch created?
- if this is a `feat` branch merging, do all satellite projects have a proper tag inside `go.mod`?

### CODE REVIEW

File: integrationTests/chainSimulator/vm/esdtImprovements_test.go [hunk 1]
```diff
@@ -0,0 +1,1449 @@
...
+	roles := [][]byte{
+		[]byte(core.ESDTRoleNFTCreate),
+		[]byte(core.ESDTRoleTransfer),
+	}
+	setAddressEsdtRoles(t, cs, addrs[0], metaESDTTokenID, roles)
+
+	log.Info("Issued metaESDT token id", "tokenID", string(metaESDTTokenID))
+
+	// issue fungible
+	fungibleTicker := []byte("FUNGIBLETICKER")
```
Comment (BeniaminDrasovean) [inline integrationTests/chainSimulator/vm/esdtImprovements_test.go:661]: Change ticker name to be <=10 and uncomment the code from L707 & L726

File: integrationTests/chainSimulator/vm/esdtImprovements_test.go [hunk 2]
```diff
@@ -0,0 +1,1449 @@
...
+
+		nonce++
+	}
+
+	err = cs.GenerateBlocks(10)
+	require.Nil(t, err)
+
+	log.Info("Step 1. check that the metaData for the NFT was saved in the user account and not on the system account")
+
+	checkMetaData(t, cs, addrs[0].Bytes, nftTokenID, nftMetaData)
```
Comment (BeniaminDrasovean) [inline integrationTests/chainSimulator/vm/esdtImprovements_test.go:745]: Check that the meta data is not saved in the system account also.

File: integrationTests/chainSimulator/vm/esdtImprovements_test.go [hunk 3]
```diff
@@ -0,0 +1,1449 @@
...
+
+	log.Info("Step 2. check that the metaData for the other token types is saved on the system account and not at the user account level")
+
+	checkMetaData(t, cs, core.SystemAccountAddress, sftTokenID, nftMetaData)
```
Comment (BeniaminDrasovean) [inline integrationTests/chainSimulator/vm/esdtImprovements_test.go:749]: Check that the meta data is not saved in the user account also. (for fungible, sft and meta)

File: integrationTests/chainSimulator/vm/esdtImprovements_test.go [hunk 4]
```diff
@@ -0,0 +1,1449 @@
...
+	nftMetaData.Nonce = []byte(hex.EncodeToString(big.NewInt(1).Bytes()))
+
+	tx = nftCreateTx(1, address.Bytes, nftTokenID, nftMetaData)
+
+	txResult, err = cs.SendTxAndGenerateBlockTilTxIsExecuted(tx, maxNumOfBlockToGenerateWhenExecutingTx)
+	require.Nil(t, err)
+	require.NotNil(t, txResult)
+	require.Equal(t, "success", txResult.Status.String())
+
+	log.Info("Call ESDTMetaDataRecreate to rewrite the meta data for the nft")
```
Comment (BeniaminDrasovean) [inline integrationTests/chainSimulator/vm/esdtImprovements_test.go:982]: ESDTMetaDataUpdate

File: integrationTests/chainSimulator/vm/esdtImprovements_test.go [hunk 5]
```diff
@@ -0,0 +1,1458 @@
...
+func TestChainSimulator_CheckTokensMetadata_TransferTokens(t *testing.T) {
+	if testing.Short() {
+		t.Skip("this is not a short test")
+	}
+
+	t.Run("transfer and check all tokens - intra shard", func(t *testing.T) {
+		transferAndCheckTokensMetaData(t, false)
+	})
+
+	t.Run("transfer and check all tokens - intra shard", func(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (ssd04) [reply integrationTests/chainSimulator/vm/esdtImprovements_test.go:63]: changed

File: integrationTests/chainSimulator/vm/esdtImprovements_test.go [hunk 6]
```diff
@@ -0,0 +1,1460 @@
...
+	roles := [][]byte{
+		[]byte(core.ESDTRoleNFTCreate),
+		[]byte(core.ESDTRoleTransfer),
+	}
+	setAddressEsdtRoles(t, cs, addrs[0], metaESDTTokenID, roles)
+
+	log.Info("Issued metaESDT token id", "tokenID", string(metaESDTTokenID))
+
+	// issue fungible
+	fungibleTicker := []byte("FUNGIBLETICKER")
```
[the remark this answers is not in the document]
Comment (ssd04) [reply integrationTests/chainSimulator/vm/esdtImprovements_test.go:133]: changed, thanks

File: integrationTests/chainSimulator/vm/esdtImprovements_test.go (hunk 1 above)
Comment (ssd04) [reply integrationTests/chainSimulator/vm/esdtImprovements_test.go:661]: done

File: integrationTests/chainSimulator/vm/esdtImprovements_test.go (hunk 4 above)
Comment (ssd04) [reply integrationTests/chainSimulator/vm/esdtImprovements_test.go:982]: done

File: integrationTests/chainSimulator/vm/esdtImprovements_test.go [hunk 7]
```diff
@@ -0,0 +1,1449 @@
...
+// 2. wait for DynamicEsdtFlag activation
+// 3. transfer the tokens to another account
+// 4. check that the metadata for all tokens is saved on the system account
+// 5. make an updateTokenID@tokenID function call on the ESDTSystem SC for all token types
+// 6. check that the metadata for all tokens is saved on the system account
+// 7. transfer the tokens to another account
+// 8. check that the metaData for the NFT was removed from the system account and moved to the user account
+// 9. check that the metaData for the rest of the tokens is still present on the system account and not on the userAccount
+// 10. do the test for both intra and cross shard txs
+func TestChainSimulator_CheckTokensMetadata_TransferTokens(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (ssd04) [reply integrationTests/chainSimulator/vm/esdtImprovements_test.go:53]: added also tests with `MultiESDTNFTTransfer`

**Review by sasurobert — APPROVED** [review]

**Review by gabi-vuls — APPROVED** [review]
Normal allin test: 1.7.12-2a6916fbce -> esdt-improvements-integrat-a3556df3a8

--- Specific errors ---

block hash does not match 1833
wrong nonce in block 772
miniblocks does not match 0
num miniblocks does not match 0
miniblock hash does not match 0
block bodies does not match 0
receipts hash missmatch 0

/------/


--- Statistics ---

Nr. of all ERRORS:  0
Nr. of all WARNS:  529
Nr. of new ERRORS:  0
Nr. of new WARNS:  0
Nr. of PANICS:  0

/------/

--- ERRORS ---

/------/

--- WARNINGS ---

/------/

**Pull request merged**